src/client/dfs/dfs_sys.c

```diff
@@ -1399,6 +1399,60 @@ dfs_sys_mkdir(dfs_sys_t *dfs_sys, const char *dir, mode_t mode,
 	return rc;
 }
 
+int
+dfs_sys_mkdir_p(dfs_sys_t *dfs_sys, const char *dir, mode_t mode, daos_oclass_id_t cid)
+{
+	int         rc      = 0;
+	int         dir_len = strnlen(dir, PATH_MAX);
+	struct stat st      = {0};
+	char       *_dir    = NULL;
+	char       *ptr     = NULL;
+
+	if (dfs_sys == NULL)
+		return EINVAL;
+	if (dir == NULL)
+		return EINVAL;
+	if (dir_len == PATH_MAX)
+		return EINVAL;
+
+	/* quick check -- if it already exists we can exit early */
+	rc = dfs_sys_stat(dfs_sys, dir, 0, &st);
```

Comment:
It's unfortunate we need this extra stat. In the future we could optimize around this by adding support to the helper functions or something to use `dfs_open_stat` to create + stat if existing, or maybe some other mechanism

Reply:
Yeah, it's only happening for the target directory though, so it seems like a minor inefficiency.

Reply:
im not sure why this is required though. you would need to recreate the parent dirs if they don't exist anyway, so you will need to parse every entry anyway. so let's say you are creating: /a/b/c/d
/a/b/c already exisits.
you are walking /a/b/c/ here, and then walking it again in the mkdir loop at the bottom.

Reply:
> Yeah, it's only happening for the target directory though, so it seems like a minor inefficiency.

it's not actually. stat will need to walk the entire path.

Reply:
OK, coming back to this... I removed the stat, but now my new unit test fails because it seems like mkdir() with an existing file doesn't fail:

```C
	/* this shouldn't work */
	rc = dfs_sys_mkdir_p(dfs_sys_mt, file, S_IWUSR | S_IRUSR, 0);
	assert_int_equal(rc, EEXIST);
```

Looking at [the comment](https://github.com/daos-stack/daos/blob/master/src/client/dfs/common.c#L632) for `create_dir()`, I'm wondering if maybe some code was lost along the way? There's no existence check, unless I'm missing something. The only thing the parent object is used for is getting the dir oclass.

Reply:
I'm seeing the same. `create_dir` just creates the object, then `dfs_mkdir` tries to insert the metadata with `DAOS_COND_DKEY_INSERT`. So if the dkey (name of the file or dir) already exists then it returns success
https://github.com/daos-stack/daos/blob/b20a41dafec49582dcdb27d917b23bccea4468db/src/client/dfs/dir.c#L58


Side note @mchaarawi what happens when `create_dir` succeeds, which does `daos_obj_open`, and then `dfs_mkdir` calls `daos_obj_close`? Does that new obj still exist? with no reference to it?

Reply:
I just restored (or added?) the existence check. If the entry exists and is a directory, then the check will return EEXIST. If it exists and is not a directory, it will return ENOTDIR. I didn't see a better way to distinguish between these states.

Reply:
the DAOS_COND_DKEY_INSERT check should return DER_EXISTS if the entry exists.
@daltonbohning objects in DAOS have no "existence" so just opening / closing an object with the obj API does not consume anything. they are just local operations on the server.

Reply:
i just verified that DFS does return EEXISTS in this case, but the new mkdir_p function coverts that to success..  i think the main issue is that we need to distinguish between EEXISTS and ENOTDIR for this function. so in that case i would suggest that an extra check is added here in case EEXIST is returned ot verify it's a dir and return ENOTDIR other wise. does that sound sensible?